### PR TITLE
discard/syncfs support on io uring

### PR DESCRIPTION
**Pull request by sulakshm**

### CODE REVIEW

File: io.c [hunk 1]
```diff
@@ -784,6 +784,87 @@ static int io_write(struct io_kiocb *req, const struct sqe_submit *s,
...
+	int ret = -EINVAL;
+	loff_t pos = s->pos;
+	size_t bytes = s->len;
+	
+	/* discard always requires a blocking context */
+	if (force_nonblock)
+		return -EAGAIN;
+
+	if (unlikely(!(req->file->f_mode & FMODE_WRITE)))
+		return -EBADF;
```
Comment (maxkozlovsky) [inline io.c:802]: EBADF does not look like the right return code here

File: io.c [hunk 2]
```diff
@@ -784,6 +784,87 @@ static int io_write(struct io_kiocb *req, const struct sqe_submit *s,
...
+	} else  if (S_ISBLK(inode->i_mode)) {
+		struct block_device *bdev = I_BDEV(inode);
+		sector_t start_sector = (pos >> SECTOR_SHIFT);
+		sector_t nsectors = sector_div(bytes, SECTOR_SIZE);
+
+		// santize byte range for sector granularity
+		if ((pos & (SECTOR_SIZE-1)) || (nsectors & (SECTOR_SIZE-1))) {
+			return -EINVAL;
+		}
+		ret = blkdev_issue_discard(bdev, start_sector, nsectors, GFP_KERNEL, 0);
```
Comment (maxkozlovsky) [inline io.c:822]: I think in 4.x+ kernels the block devices support fallocate. uring is supported from 4.15, so you might be able to simply call fallocate all the time.

File: io.c [hunk 3]
```diff
@@ -784,6 +784,87 @@ static int io_write(struct io_kiocb *req, const struct sqe_submit *s,
...
+	} else {
+		return -EINVAL;
+	}
+
+	// only for success do the below, as its done through the wq caller context
+	if (!req) {
```
Comment (maxkozlovsky) [inline io.c:828]: if (!ret)

File: io.c [hunk 4]
```diff
@@ -784,6 +784,87 @@ static int io_write(struct io_kiocb *req, const struct sqe_submit *s,
...
+		up_read(&sb->s_umount);
+	} else if (S_ISBLK(inode->i_mode)) {
+		struct block_device *bdev = I_BDEV(inode);
+		ret = blkdev_issue_flush(bdev, GFP_KERNEL, NULL);
+	} else {
+		return -EOPNOTSUPP;
+	}
+
+	// only for success do the below, as its done through the wq caller context
+	if (!req) {
```
Comment (maxkozlovsky) [inline io.c:860]: if (!ret)

File: io.c (hunk 3 above)
Comment (sulakshm) [reply io.c]: fixed, thanks,

**Pull request merged**